### DISCUSSION TITLE
How to use arguments to Smart.Format

### DISCUSSION THREAD

**Discussion by Bluscream** [axuno/SmartFormat#352] [Q&A] [answered]
Hi! Thanks for reporting this bug! 

Please fill in the relevant information from this template so that we can help you best.

**Smart.Format version**: 3.2.2

**Framework version**:  (e.g. net5.0) net6

- Please provide a working source code example to reproduce the bug
https://github.com/Bluscream/battlebitapirunner-modules/blob/f706b53ef15d943ddcc7697ec566e90c1e16b4b9/modules/Logger.cs#L118

- What is the current result?
Crash
- What is the expected result?
formatted DateTime.Now
- Please post full exception details if applicable (message, stacktrace, inner exceptions)
```
Unhandled exception. SmartFormat.Core.Formatting.FormattingException: Error parsing format string: No source extension could handle the selector named "now" at 2
[{now}] API Modules loaded
--^
   at SmartFormat.SmartFormatter.FormatError(FormatItem errorItem, Exception innerException, Int32 startIndex, IFormattingInfo formattingInfo)
   at SmartFormat.SmartFormatter.Format(FormattingInfo formattingInfo)
   at SmartFormat.SmartFormatter.Format(FormatDetails formatDetails, Format format, Object current)
   at SmartFormat.SmartFormatter.Format(IFormatProvider provider, String format, IList`1 args)
   at LoggerBattlebitModule.Logger.FormatString(String input, Object[] parms) in ./modules\Logger.cs:line 118
   at LoggerBattlebitModule.Logger.HandleEvent(LogConfigurationEntry config, Object[] parms) in ./modules\Logger.cs:line 177
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_1(Object state)
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
NativeCommandExitException: Program "dotnet.exe" ended with non-zero exit code: -532462766.
```
- Did you find a workaround? yes/no
no
- Is there a version in which it worked? 
not that i know of
- Can you help us by writing an unit test?
what?

**Comment by axunonb** [comment]
The provided link is not a working code sample, as it is missing the format string and the arguments to `Smart.Format()`. Also it's unclear which extensions are in use.

**Reply by Bluscream** [reply]
The strings to format are provided via the config class which is easily observed within the code:
https://github.com/Bluscream/battlebitapirunner-modules/blob/master/modules/Logger.cs#L344-L348

Loaded dependencies are:
![](https://i.imgur.com/xjh9gYP.png)

**Reply by axunonb** [reply]
The question was not about loaded dependencies, but about Smart.Format extensions that are loaded.

**Reply by Bluscream** [reply]
The dependecy folder contains every library that's loaded for the module to use at compile/runtime. And the usings are in the module code. Am i missing something?

**Comment by axunonb** [comment, marked as answer]
```CSharp
var player = new { Name = "Joe", SteamID = 1234, IP = "127.0.0.1" };
var now =  DateTime.Now.ToString(CultureInfo.InvariantCulture);
var arg = new { player, now };
var format = "[{now}] [-] {player.Name} ({player.SteamID})) [{player.IP}]";
var result = Smart.Format(format, arg);
// "[09/02/2023 08:58:07] [-] Joe (1234)) [127.0.0.1]"
//
// or
//
format = "[{0}] [-] {1.Name} ({1.SteamID})) [{1.IP}]";
result = Smart.Format(format, now, player);
// [09/02/2023 09:02:01] [-] Joe (1234)) [127.0.0.1]
```

**Reply by Bluscream** [reply]
Isn't that exactly what i do with `params`?

Look at this abomination; it works, but at what cost https://github.com/Bluscream/battlebitapirunner-modules/blob/00ddf4966428b4c5489fe85a07ccedae1dcba3e5/modules/Logger.cs#L175

**Reply by axunonb** [reply]
> Isn't that exactly what i do with params?

Similar, yes. The way you're using the placeholder `{now}` doesn't combine with the given parameter.
To understand the difference please have a glance at https://github.com/axuno/SmartFormat/wiki/Syntax%2C-Terminology
